openassessment/workflow_batch_update_api.py

```diff
@@ -0,0 +1,629 @@
+    """
+    Verify on ORA and Course level if flexible peer grading is "ON"
+
+    Args:
+        ora_block (OpenAssessmentBlock): ORA block
+        course_block (CourseBlock): Course block
+
+    Returns:
+        bool: True if given ORA is configured with flexible peer grading set "ON"
+    """
+    if ora_block is not None:
+        workflow_requirements = ora_block.workflow_requirements()
+        if workflow_requirements.get('peer') and workflow_requirements['peer'].get('enable_flexible_grading'):
+            return True
+    if course_block is not None:
+        return course_block.force_on_flexible_peer_openassessments
+
+    return False
+
+
+def get_blocked_peer_workflows(course_id=None, item_id=None, submission_uuid=None):
+    """
+    Retrieve ORA peer workflows not completed for >7 days
+
+    Returns:
+        list (PeerWorkflow): list of workfows not completed for > 7 days
+    """
+
+    filters = {
+        'created_at__lte': timezone.now() - datetime.timedelta(days=7),
+        'completed_at__isnull': True,
+    }
+    if course_id is not None:
+        filters['course_id'] = course_id
+    if item_id is not None:
+        filters['item_id'] = item_id
+    if submission_uuid is not None:
+        filters['submission_uuid'] = item_id
+
+    return PeerWorkflow.objects.filter(**filters)
+
+
+def get_workflow_update_parameters(peer_workflows):
+    """
+    Generates dictionary containing data required to update ORA workflows for different scopes.
+    This data structure is used as a local cache to avoid redundant DB queries during
+    batch workflow update process
+
+    Structure:
+    ```
+    {
+      "courses": [
+        {
+          "course_id": "value",
+          "course_settings": {},
+          "assessments": [
+            {
+              "item_id": "value",
+              "assessment_requirements": {},
+              "submissions": [
+                {
+                  "submission_uuid": "value"
+                }
+              ]
+            }
+          ]
+        }
+      ]
+    }
+    ```
+    """
+    workflow_update_parameters_dict = {}
+    store = modulestore()
+    for peer_workflow in peer_workflows:
+        try:
+            # retrieve openassessment block
+            ora_block_key = UsageKey.from_string(peer_workflow.item_id)
+            ora_block = store.get_item(ora_block_key)
+
+            # retrieve course block
+            course_block_key = UsageKey.from_string(peer_workflow.course_id)
+            course_block = store.get_item(course_block_key)
+
+            if is_flexible_peer_grading_on(ora_block, course_block):
+                # we are interested only in submissions for ORA with flexible peer grading configured
+                course_settings = {
+                    'force_on_flexible_peer_openassessments': course_block.force_on_flexible_peer_openassessments}
+                _add_workflow_update_parameters(workflow_update_parameters_dict,
```

Comment:
Nit: I personally dislike this style of function (pass in a dict which gets modified inside the function) as it hides and can tweak data in ways that are hard to inspect and test for. Would prefer:

```suggestion
                update_params = get_update_params_for_submission(<whole lotta args, for some reason>)
                workflow_update_parameters_dict.update(update_params)
```

This will also let you test that data transform more easily, if you needed to test it.

Reply:
Agreed on updating via reference, typically you would additionally return updated object (like in your example), however, for the internal/private/helper call this is acceptable IMO